platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

```diff
@@ -0,0 +1,350 @@
+
+### **Background**
+
+The current content-build for va.gov runs once an hour and generates all HTML pages. The current build takes approximatly 15 minutes for page to show up on va.gov. When the amount of content in Drupal reaches 48K nodes, the build will take close to 45 minutes to publish content to va.gov.  Editors are often confused about when the content they published in the CMS will be live on va.gov.
+
+Editors coming from Team Sites expect publication of content within 5 minutes.  As editors migrate from Team Sits to Drupal, editors often do not understand why the content they publish does not show up on va.gov in a timely mannter.
+
+
+### **High Level Design**
+
+The same code used to generate the HTML for the preview server can be used to generate the HTML for va.gov.  
+
+The following diagram shows a high level design of how a node in Drupal would be published “immediately” on va.gov.
+
+Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an S3 bucket which serves the html for va.gov.  
+
+The code used by the lambda to generate the HTML will utilize the current preview server code.  As part of this work, the preview server code will be abstracted and improved to provide production ready HTML.
+
+![image](https://user-images.githubusercontent.com/121603/121699098-a2cd3b00-ca9c-11eb-8a0e-5e15cce7ce4d.png)
+
+
+Editable link: [https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#](https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#)
+
+
+## **Specifics**
+
+
+### **Detailed Design**
+
+
+#### Drupal Triggers Content Publishing
+
+When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  Drupal will also have logic to not allow a Small Batch Publishing if a full build is in progress.
+
+Drupal will track the queue progress using the [advancedqueue](https://www.drupal.org/project/advancedqueue) and the Public Batch API described below. The advancedqueue module allows for observability and reporting.
+
+
+#### Preview Server Changes
```

Comment:
Just food for thought that I think engineers use the preview functionality locally as well, so another point of QA

Reply:
I added notes about validating local build.